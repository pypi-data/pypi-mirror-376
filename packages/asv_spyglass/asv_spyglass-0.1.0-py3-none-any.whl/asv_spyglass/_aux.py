# Kanged from rgpycrumbs
def getstrform(pathobj):
    return str(pathobj.absolute())
