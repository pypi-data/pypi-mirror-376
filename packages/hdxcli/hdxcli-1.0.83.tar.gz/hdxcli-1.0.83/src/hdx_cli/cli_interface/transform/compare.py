import copy
from pathlib import Path

import click

from hdx_cli.auth.context_builder import load_user_context
from hdx_cli.cli_interface.common.click_extensions import HdxCommand
from hdx_cli.cli_interface.common.undecorated_click_commands import basic_transform
from hdx_cli.cli_interface.transform.comparator.engine import TransformComparator
from hdx_cli.config.paths import PROFILE_CONFIG_FILE
from hdx_cli.library_api.common.logging import get_logger
from hdx_cli.library_api.utility.decorators import report_error_and_exit
from hdx_cli.library_api.utility.file_handling import read_json_from_file
from hdx_cli.models import ProfileUserContext, ProfileLoadContext

logger = get_logger()


def _fetch_transform_from_cluster(
    ctx: click.Context,
    project_name: str,
    table_name: str,
    transform_name: str
) -> dict:
    """Fetches a single transform's configuration from
    the cluster. Returns the configuration as a dictionary."""
    user_profile = ctx.parent.obj["usercontext"]
    ProfileUserContext.update_context(
        user_profile, projectname=project_name, tablename=table_name, transformname=transform_name
    )
    basic_transform(ctx)
    return ctx.obj.get("specific_resource")


def _load_transform_spec(
    ctx: click.Context,
    specifier: str,
) -> tuple[dict, str]:
    """Loads a transform from a local file path or a cluster reference.
    A cluster reference is in the format 'project.table.transform'.
    Returns a tuple of (transform_data, description)."""
    if Path(specifier).is_file() and specifier.endswith('.json'):
        logger.debug(f"Loading transform from local file: {specifier}")
        return read_json_from_file(specifier), f"local file: {specifier}"

    # Check for cluster reference format: project.table.transform
    parts = specifier.split('.')
    if len(parts) == 3:
        project, table, transform = parts
        user_profile = ctx.parent.obj["usercontext"]
        logger.debug(f"Fetching transform '{transform}' from cluster '{user_profile.profilename}'...")
        data = _fetch_transform_from_cluster(ctx, project, table, transform)
        return data, f"cluster '{user_profile.profilename}': {specifier}"

    raise click.BadParameter(
        f"Invalid format for '{specifier}'. Provide a path to a .json file or a "
        "cluster reference like 'project_name.table_name.transform_name'."
    )


@click.command(cls=HdxCommand, name="compare")
@click.argument("transform_a_spec", metavar="TRANSFORM_A")
@click.argument("transform_b_spec", metavar="TRANSFORM_B")
@click.option(
    "--profile-b",
    help="Profile for fetching TRANSFORM_B, if different from the default.",
    metavar="PROFILE_NAME",
)
@click.pass_context
@report_error_and_exit(exctype=Exception)
def compare(
    ctx_a: click.Context,
    transform_a_spec: str,
    transform_b_spec: str,
    profile_b: str
):
    """Compares two transforms, showing differences in their settings.

    Transforms can be specified as a local JSON file path or as a reference
    to a transform on a cluster in the format: `project_name.table_name.transform_name`.

    When comparing a transform on a different cluster, use `--profile-b` to
    specify the profile details for `TRANSFORM_B`.

    \b
    Examples:
      # Compare two local JSON files
      hdxcli transform compare path/to/transform_A.json path/to/transform_B.json

    \b
      # Compare a local file to a transform on the default profile
      hdxcli transform compare path/to/transform_A.json my_project.my_table.my_transform

    \b
      # Compare two transforms on the same 'default' profile
      hdxcli transform compare proj_A.tbl_A.tf_A proj_B.tbl_B.tf_B

    \b
      # Compare a transform on the 'dev' profile to one on another (profile 'prod')
      hdxcli --profile dev transform compare p1.t1.tf1 p2.t2.tf2 --profile-b prod
    """
    ctx_b = copy.deepcopy(ctx_a)
    # Determine ctx for TRANSFORM_B
    if profile_b:
        load_context = ProfileLoadContext(name=profile_b, config_file=PROFILE_CONFIG_FILE)
        profile_for_b = load_user_context(load_context)
        ctx_b.parent.obj["'profilecontext'"] = load_context
        ctx_b.parent.obj["usercontext"] = profile_for_b

    # Load both transforms
    data_a, desc_a = _load_transform_spec(ctx_a, transform_a_spec)
    data_b, desc_b = _load_transform_spec(ctx_b, transform_b_spec)

    comparator = TransformComparator(data_a, data_b, desc_a, desc_b)
    comparator.run()
    comparator.display()
