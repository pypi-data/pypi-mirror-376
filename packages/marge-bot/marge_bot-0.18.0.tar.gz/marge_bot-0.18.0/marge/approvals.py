from typing import TYPE_CHECKING, Any, Union, cast

from . import gitlab

if TYPE_CHECKING:
    from . import merge_request as mb_merge_request


class Approvals(gitlab.Resource):
    """Approval info for a MergeRequest.
    https://docs.gitlab.com/ee/api/merge_request_approvals.html"""

    def refetch_info(self) -> None:
        approver_url = (
            f"/projects/{self.project_id}/merge_requests/{self.iid}/approvals"
        )
        info = self._api.call(gitlab.GET(approver_url))

        if TYPE_CHECKING:
            assert isinstance(info, dict)
        self._info = info

    @property
    def id(self) -> int:
        raise NotImplementedError()

    @property
    def iid(self) -> int:
        return cast(int, self.info["iid"])

    @property
    def project_id(self) -> int:
        return cast(int, self.info["project_id"])

    @property
    def approvals_left(self) -> int:
        return cast(int, self.info.get("approvals_left", 0))

    @property
    def sufficient(self) -> bool:
        return not self.approvals_left

    @property
    def approver_usernames(self) -> list[str]:
        return [who["user"]["username"] for who in self.info["approved_by"]]

    @property
    def approver_ids(self) -> list[int]:
        """Return the uids of the approvers."""
        return [who["user"]["id"] for who in self.info["approved_by"]]

    def reapprove(self) -> None:
        """Impersonates the approvers and re-approves the merge_request as them.

        The idea is that we want to get the approvers, push the rebased branch
        (which may invalidate approvals, depending on GitLab settings) and then
        restore the approval status.
        """
        self.approve(self)

    def approve(self, obj: Union["Approvals", "mb_merge_request.MergeRequest"]) -> None:
        """Approve an object which can be a merge_request or an approval."""
        approve_url = f"/projects/{obj.project_id}/merge_requests/{obj.iid}/approve"

        for uid in self.approver_ids:
            self._api.call(gitlab.POST(approve_url), sudo=uid)


class CustomApprovals(Approvals):
    """Allows a limited way to manage approvals on CE."""

    def __init__(
        self,
        api: gitlab.Api,
        info: dict[str, Any],
        allowed_approvers: list[str],
        approvals_required: int = 1,
    ):
        super().__init__(api, info)
        self._allowed_approvers = allowed_approvers
        self._approvals_required = approvals_required

    @property
    def allowed_approvals_usernames(self) -> list[str]:
        """Filter approver_usernames to only those allowed."""
        return [
            username
            for username in self.approver_usernames
            if username in self._allowed_approvers
        ]

    @property
    def approvals_left(self) -> int:
        return max(0, self._approvals_required - len(self.allowed_approvals_usernames))

    @property
    def sufficient(self) -> bool:
        return not self.approvals_left
