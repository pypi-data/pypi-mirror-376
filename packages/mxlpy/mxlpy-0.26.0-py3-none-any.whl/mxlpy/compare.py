"""Docstring."""

from __future__ import annotations

from dataclasses import dataclass
from typing import TYPE_CHECKING, cast

import pandas as pd
from wadler_lindig import pformat

from mxlpy import plot
from mxlpy.simulator import Simulator
from mxlpy.types import Result, unwrap

if TYPE_CHECKING:
    from mxlpy.model import Model
    from mxlpy.types import ArrayLike

__all__ = [
    "ProtocolComparison",
    "SteadyStateComparison",
    "TimeCourseComparison",
    "protocol_time_courses",
    "steady_states",
    "time_courses",
]


@dataclass
class SteadyStateComparison:
    """Compare two steady states."""

    res1: Result
    res2: Result

    def __repr__(self) -> str:
        """Return default representation."""
        return pformat(self)

    @property
    def variables(self) -> pd.DataFrame:
        """Compare the steady state variables."""
        ss1 = self.res1.get_variables().iloc[-1]
        ss2 = self.res2.get_variables().iloc[-1]
        diff = ss2 - ss1
        return pd.DataFrame(
            {"m1": ss1, "m2": ss2, "diff": diff, "rel_diff": diff / ss1}
        )

    @property
    def fluxes(self) -> pd.DataFrame:
        """Compare the steady state fluxes."""
        ss1 = self.res1.get_fluxes().iloc[-1]
        ss2 = self.res2.get_fluxes().iloc[-1]
        diff = ss2 - ss1
        return pd.DataFrame(
            {"m1": ss1, "m2": ss2, "diff": diff, "rel_diff": diff / ss1}
        )

    @property
    def all(self) -> pd.DataFrame:
        """Compare both steady-state variables and fluxes."""
        ss1 = self.res1.get_combined().iloc[-1]
        ss2 = self.res2.get_combined().iloc[-1]
        diff = ss2 - ss1
        return pd.DataFrame(
            {"m1": ss1, "m2": ss2, "diff": diff, "rel_diff": diff / ss1}
        )

    def plot_variables(self, title: str = "Variables") -> plot.FigAxs:
        """Plot the relative difference of steady-state variables."""
        fig, axs = plot.bars_autogrouped(self.variables["rel_diff"], ylabel="")
        plot.grid_labels(axs, ylabel="Relative difference")
        fig.suptitle(title)
        return fig, axs

    def plot_fluxes(self, title: str = "Fluxes") -> plot.FigAxs:
        """Plot the relative difference of steady-state fluxes."""
        fig, axs = plot.bars_autogrouped(self.fluxes["rel_diff"], ylabel="")
        plot.grid_labels(axs, ylabel="Relative difference")
        fig.suptitle(title)
        return fig, axs

    def plot_all(self, title: str = "Variables and Fluxes") -> plot.FigAxs:
        """Plot the relative difference of steady-state variables and fluxes."""
        combined = self.all

        fig, axs = plot.bars_autogrouped(combined["rel_diff"], ylabel="")
        plot.grid_labels(axs, ylabel="Relative difference")
        fig.suptitle(title)
        return fig, axs


@dataclass
class TimeCourseComparison:
    """Compare two time courses."""

    res1: Result
    res2: Result

    def __repr__(self) -> str:
        """Return default representation."""
        return pformat(self)

    # @property
    # def variables(self) -> pd.DataFrame:
    #     """Compare the steady state variables."""
    #     ss1 = self.res1.get_variables()
    #     ss2 = self.res2.get_variables()
    #     diff = ss2 - ss1
    #     return pd.DataFrame(
    #         {"m1": ss1, "m2": ss2, "diff": diff, "rel_diff": diff / ss1}
    #     )

    # @property
    # def fluxes(self) -> pd.DataFrame:
    #     """Compare the steady state fluxes."""
    #     ss1 = self.res1.get_fluxes()
    #     ss2 = self.res2.get_fluxes()
    #     diff = ss2 - ss1
    #     return pd.DataFrame(
    #         {"m1": ss1, "m2": ss2, "diff": diff, "rel_diff": diff / ss1}
    #     )

    def plot_variables_relative_difference(self) -> plot.FigAxs:
        """Plot the relative difference of time course variables."""
        c1 = self.res1.variables
        c2 = self.res2.variables

        rel_diff = ((c2.loc[:, cast(list[str], c1.columns)] - c1) / c1).fillna(0)
        fig, axs = plot.line_autogrouped(rel_diff, ylabel="")
        plot.grid_labels(axs, ylabel="Relative difference")
        fig.suptitle("Variables")
        return fig, axs

    def plot_fluxes_relative_difference(self) -> plot.FigAxs:
        """Plot the relative difference of time course fluxes."""
        v1 = self.res1.fluxes
        v2 = self.res2.fluxes

        rel_diff = ((v2.loc[:, cast(list[str], v1.columns)] - v1) / v1).fillna(0)
        fig, axs = plot.line_autogrouped(rel_diff, ylabel="")
        plot.grid_labels(axs, ylabel="Relative difference")
        fig.suptitle("Fluxes")
        return fig, axs


@dataclass
class ProtocolComparison:
    """Compare two protocol time courses."""

    res1: Result
    res2: Result
    protocol: pd.DataFrame

    def __repr__(self) -> str:
        """Return default representation."""
        return pformat(self)

    # @property
    # def variables(self) -> pd.DataFrame:
    #     """Compare the steady state variables."""
    #     ss1 = self.res1.get_variables()
    #     ss2 = self.res2.get_variables()
    #     diff = ss2 - ss1
    #     return pd.DataFrame(
    #         {"m1": ss1, "m2": ss2, "diff": diff, "rel_diff": diff / ss1}
    #     )

    # @property
    # def fluxes(self) -> pd.DataFrame:
    #     """Compare the steady state fluxes."""
    #     ss1 = self.res1.get_fluxes()
    #     ss2 = self.res2.get_fluxes()
    #     diff = ss2 - ss1
    #     return pd.DataFrame(
    #         {"m1": ss1, "m2": ss2, "diff": diff, "rel_diff": diff / ss1}
    #     )

    def plot_variables_relative_difference(
        self,
        shade_protocol_variable: str | None = None,
    ) -> plot.FigAxs:
        """Plot the relative difference of time course variables."""
        c1 = self.res1.variables
        c2 = self.res2.variables

        rel_diff = ((c2.loc[:, cast(list[str], c1.columns)] - c1) / c1).fillna(0)
        fig, axs = plot.line_autogrouped(rel_diff, ylabel="")
        plot.grid_labels(axs, ylabel="Relative difference")
        fig.suptitle("Variables")

        if shade_protocol_variable is not None:
            protocol = self.protocol[shade_protocol_variable]
            for ax in axs:
                plot.shade_protocol(protocol=protocol, ax=ax)
        return fig, axs

    def plot_fluxes_relative_difference(
        self,
        shade_protocol_variable: str | None = None,
    ) -> plot.FigAxs:
        """Plot the relative difference of time course fluxes."""
        v1 = self.res1.fluxes
        v2 = self.res2.fluxes

        rel_diff = ((v2.loc[:, cast(list[str], v1.columns)] - v1) / v1).fillna(0)
        fig, axs = plot.line_autogrouped(rel_diff, ylabel="")
        plot.grid_labels(axs, ylabel="Relative difference")
        fig.suptitle("Fluxes")

        if shade_protocol_variable is not None:
            protocol = self.protocol[shade_protocol_variable]
            for ax in axs:
                plot.shade_protocol(protocol=protocol, ax=ax)
        return fig, axs


def steady_states(m1: Model, m2: Model) -> SteadyStateComparison:
    """Compare the steady states of two models."""
    return SteadyStateComparison(
        res1=unwrap(Simulator(m1).simulate_to_steady_state().get_result()),
        res2=unwrap(Simulator(m2).simulate_to_steady_state().get_result()),
    )


def time_courses(m1: Model, m2: Model, time_points: ArrayLike) -> TimeCourseComparison:
    """Compare the time courses of two models."""
    return TimeCourseComparison(
        res1=unwrap(
            Simulator(m1).simulate_time_course(time_points=time_points).get_result()
        ),
        res2=unwrap(
            Simulator(m2).simulate_time_course(time_points=time_points).get_result()
        ),
    )


def protocol_time_courses(
    m1: Model,
    m2: Model,
    protocol: pd.DataFrame,
) -> ProtocolComparison:
    """Compare the time courses of two models."""
    return ProtocolComparison(
        res1=unwrap(Simulator(m1).simulate_protocol(protocol=protocol).get_result()),
        res2=unwrap(Simulator(m2).simulate_protocol(protocol=protocol).get_result()),
        protocol=protocol,
    )
