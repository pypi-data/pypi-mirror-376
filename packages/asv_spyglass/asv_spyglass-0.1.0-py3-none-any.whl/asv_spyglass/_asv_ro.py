import itertools
import re
from pathlib import Path

from asv.util import load_json as asv_json_load

from asv_spyglass._aux import getstrform


class ReadOnlyASVBenchmarks:
    """Read-only holder for a set of ASV benchmarks."""

    api_version = 2

    def __init__(self, benchmarks_file: Path, regex: str | list[str] = None):
        """
        Initialize and load benchmarks from a JSON file, optionally filtering them.

        Args:
            benchmarks_file (Path): Path to the benchmarks JSON file.
            regex (Union[str, list[str]], optional): Regular expression(s) to filter benchmarks.
                Defaults to None (all benchmarks included).
        """
        d = asv_json_load(getstrform(benchmarks_file), api_version=self.api_version)
        self._base_benchmarks = {}  # Store all benchmarks here
        self._benchmark_selection = {}  # Track selected parameter combinations
        self.filtered_benchmarks = {}  # Store selected benchmarks here after parameter expansion

        if not regex:
            regex = []
        if isinstance(regex, str):
            regex = [regex]

        for benchmark in d.values():
            self._base_benchmarks[benchmark["name"]] = benchmark
            if benchmark.get("params"):
                self._benchmark_selection[benchmark["name"]] = []
                param_tuples = [tuple(map(str, p)) for p in benchmark["params"]]
                for idx, param_set in enumerate(itertools.product(*param_tuples)):
                    name = f"{benchmark['name']}({', '.join(param_set)})"
                    if not regex or any(re.search(reg, name) for reg in regex):
                        self.filtered_benchmarks[name] = (
                            benchmark  # Store with full name
                        )
                        self._benchmark_selection[benchmark["name"]].append(idx)
            else:
                self._benchmark_selection[benchmark["name"]] = None
                if not regex or any(re.search(reg, benchmark["name"]) for reg in regex):
                    self.filtered_benchmarks[benchmark["name"]] = benchmark

    def __repr__(self) -> str:
        """Return a string representation of the filtered benchmarks."""
        import pprint

        pp = pprint.PrettyPrinter()
        return pp.pformat(self.filtered_benchmarks)

    @property
    def benchmarks(self) -> dict[str, dict]:
        """Get a dictionary of filtered benchmarks."""
        return self.filtered_benchmarks
