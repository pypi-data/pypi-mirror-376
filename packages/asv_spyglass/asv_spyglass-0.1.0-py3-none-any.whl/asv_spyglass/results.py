import dataclasses
import re
from collections import namedtuple

import polars as pl
from asv import results

from asv_spyglass._asv_ro import ReadOnlyASVBenchmarks

ASVResult = namedtuple(
    "ASVResult",
    [
        "key",
        "params",
        "value",
        "stats",
        "samples",
        "version",
        "machine",
        "env_name",
    ],
)


def result_iter(bdot):
    for key in bdot.get_all_result_keys():
        params = bdot.get_result_params(key)
        result_value = bdot.get_result_value(key, params)
        result_stats = bdot.get_result_stats(key, params)
        result_samples = bdot.get_result_samples(key, params)
        result_version = bdot.benchmark_version.get(key)
        yield ASVResult(
            key,
            params,
            result_value,
            result_stats,
            result_samples,
            result_version,
            bdot.params["machine"],
            bdot.env_name,
        )


@dataclasses.dataclass
class PreparedResult:
    """Augmented with information from the benchmarks.json"""

    units: dict
    results: dict
    stats: dict
    versions: dict
    machine_name: str
    env_name: str
    param_names: list

    def __iter__(self):
        for field in dataclasses.fields(self):
            yield getattr(self, field.name)

    def to_df(self):
        """
        Converts a PreparedResult object to a DataFrame, exploding the
        parameters in the keys and flattening the results.
        """

        data = []
        for key, result in self.results.items():
            # Extract benchmark name and parameters
            match = re.match(r"(.+)\((.*)\)", key)
            if match:
                benchmark_name = match.group(1)
                params = match.group(2).split(", ")
            else:
                benchmark_name = key
                params = []

            # Flatten the results tuple
            stats_dict, samples = self.stats[key]
            if stats_dict is None:
                stats_dict_maybe_null = {
                    "ci_99_a": None,
                    "ci_99_b": None,
                    "q_25": None,
                    "q_75": None,
                    "number": None,
                    "repeat": None,
                }
            else:
                stats_dict_maybe_null = stats_dict

            row = {
                "benchmark_base": benchmark_name,
                "name": key,
                "result": result,
                "units": self.units[key],
                "machine": self.machine_name,
                "env": self.env_name,
                "version": self.versions[key],
                **stats_dict_maybe_null,  # Expand the stats dictionary
                "samples": samples,
            }
            # Combine param and parameter name for column names
            row.update(
                dict(
                    zip(
                        [f"param_{name}" for name in self.param_names[key]],
                        params,
                    )
                )
            )
            data.append(row)

        return pl.DataFrame(data)
