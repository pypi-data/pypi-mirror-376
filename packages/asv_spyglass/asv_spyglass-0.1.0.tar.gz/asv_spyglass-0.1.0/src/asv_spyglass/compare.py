import math
from pathlib import Path

import polars as pl
import tabulate
from asv import results
from asv.commands.compare import _is_result_better, _isna, unroll_result
from asv.console import log
from asv.util import human_value
from asv_runner.console import color_print
from asv_runner.statistics import get_err

from asv_spyglass._asv_ro import ReadOnlyASVBenchmarks
from asv_spyglass.results import PreparedResult, result_iter


class ResultPreparer:
    """
    Prepares benchmark results for comparison by extracting relevant data
    like units, values, stats, versions, and machine/environment names.
    """

    def __init__(self, benchmarks):
        """
        Initializes ResultPreparer with benchmark data.

        Args:
            benchmarks: Benchmark data used for extracting units.
        """
        self.benchmarks = benchmarks

    def prepare(self, result_data):
        """
        Processes result data and returns extracted information.

        Args:
            result_data: Result data to be processed.

        Returns:
            tuple: A tuple containing units, results, stats, versions,
                   and the machine/environment name.
        """
        units = {}
        results = {}
        ss = {}
        versions = {}
        param_names = {}
        btype = {}
        machine_env_name = None

        for (
            key,
            params,
            value,
            stats,
            samples,
            version,
            machine,
            env_name,
        ) in result_iter(result_data):
            machine_env_name = f"{machine}/{env_name}"
            for name, value, stats, samples in unroll_result(
                key, params, value, stats, samples
            ):
                results[name] = value
                # TODO(hz): Split samples out later when _is_result_better is
                # not used anymore, currently in a tuple for ASV compatibility
                ss[name] = (stats, samples)
                versions[name] = version
                # HACK(hz): The names already include the parameters i.e.
                # benchmark(param) so this works around the situation
                bench_key = [x for x in self.benchmarks.keys() if key in x][0]
                units[name] = self.benchmarks.get(bench_key, {}).get("unit")
                btype[name] = self.benchmarks.get(bench_key, {}).get("type")
                param_names[name] = self.benchmarks.get(bench_key, {}).get(
                    "param_names"
                )

        machine, env_name = machine_env_name.split("/")
        return PreparedResult(
            units=units,
            results=results,
            stats=ss,
            versions=versions,
            machine_name=machine,
            env_name=env_name,
            param_names=param_names,
        )


def do_compare(
    b1,
    b2,
    bdat,
    factor=1.1,
    split=False,
    only_changed=False,
    sort="default",
    machine=None,
    env_spec=None,
    use_stats=True,
):
    # Load results
    res_1 = results.Results.load(b1)
    res_2 = results.Results.load(b2)

    # Initialize benchmarks
    benchmarks = ReadOnlyASVBenchmarks(Path(bdat)).benchmarks

    # Prepare results using the ResultPreparer class
    preparer = ResultPreparer(benchmarks)
    prepared_results_1 = preparer.prepare(res_1)
    prepared_results_2 = preparer.prepare(res_2)
    # Kanged from compare.py

    # Extract data from prepared results
    results_1 = prepared_results_1.results
    results_2 = prepared_results_2.results
    ss_1 = prepared_results_1.stats
    ss_2 = prepared_results_2.stats
    versions_1 = prepared_results_1.versions
    versions_2 = prepared_results_2.versions
    units = prepared_results_1.units

    machine_env_names = set()
    mname_1 = f"{prepared_results_1.machine_name}/{prepared_results_1.env_name}"
    mname_2 = f"{prepared_results_2.machine_name}/{prepared_results_2.env_name}"
    machine_env_names.add(mname_1)
    machine_env_names.add(mname_2)

    benchmarks_1 = set(results_1.keys())
    benchmarks_2 = set(results_2.keys())
    joint_benchmarks = sorted(list(benchmarks_1 | benchmarks_2))
    bench = {}

    if split:
        bench["green"] = []
        bench["red"] = []
        bench["lightgrey"] = []
        bench["default"] = []
    else:
        bench["all"] = []

    worsened = False
    improved = False

    for benchmark in joint_benchmarks:
        if benchmark in results_1:
            time_1 = results_1[benchmark]
        else:
            time_1 = math.nan

        if benchmark in results_2:
            time_2 = results_2[benchmark]
        else:
            time_2 = math.nan

        if benchmark in ss_1 and ss_1[benchmark][0]:
            err_1 = get_err(time_1, ss_1[benchmark][0])
        else:
            err_1 = None

        if benchmark in ss_2 and ss_2[benchmark][0]:
            err_2 = get_err(time_2, ss_2[benchmark][0])
        else:
            err_2 = None

        version_1 = versions_1.get(benchmark)
        version_2 = versions_2.get(benchmark)

        if _isna(time_1) or _isna(time_2):
            ratio = "n/a"
            ratio_num = 1e9
        else:
            try:
                ratio_num = time_2 / time_1
                ratio = f"{ratio_num:6.2f}"
            except ZeroDivisionError:
                ratio_num = 1e9
                ratio = "n/a"

        if version_1 is not None and version_2 is not None and version_1 != version_2:
            # not comparable
            color = "lightgrey"
            mark = "x"
        elif time_1 is not None and time_2 is None:
            # introduced a failure
            color = "red"
            mark = "!"
            worsened = True
        elif time_1 is None and time_2 is not None:
            # fixed a failure
            color = "green"
            mark = " "
            improved = True
        elif time_1 is None and time_2 is None:
            # both failed
            color = "default"
            mark = " "
        elif _isna(time_1) or _isna(time_2):
            # either one was skipped
            color = "default"
            mark = " "
        elif _is_result_better(
            time_2,
            time_1,
            ss_2.get(benchmark),
            ss_1.get(benchmark),
            factor,
            use_stats=use_stats,
        ):
            color = "green"
            mark = "-"
            improved = True
        elif _is_result_better(
            time_1,
            time_2,
            ss_1.get(benchmark),
            ss_2.get(benchmark),
            factor,
            use_stats=use_stats,
        ):
            color = "red"
            mark = "+"
            worsened = True
        else:
            color = "default"
            mark = " "

            # Mark statistically insignificant results
            if _is_result_better(
                time_1, time_2, None, None, factor
            ) or _is_result_better(time_2, time_1, None, None, factor):
                ratio = "~" + ratio.strip()

        if only_changed and mark in (" ", "x"):
            continue

        unit = units[benchmark]

        details = "{0:1s} {1:>15s}  {2:>15s} {3:>8s}  ".format(
            mark,
            human_value(time_1, unit, err=err_1),
            human_value(time_2, unit, err=err_2),
            ratio,
        )
        split_line = details.split()
        if len(machine_env_names) > 1:
            benchmark_name = f"{benchmark} [{mname_1} -> {mname_2}]"
        else:
            benchmark_name = benchmark
        if len(split_line) == 4:
            split_line += [benchmark_name]
        else:
            split_line = [" "] + split_line + [benchmark_name]
        if split:
            bench[color].append(split_line)
        else:
            bench["all"].append(split_line)

    if split:
        keys = ["green", "default", "red", "lightgrey"]
    else:
        keys = ["all"]

    titles = {}
    titles["green"] = "Benchmarks that have improved:"
    titles["default"] = "Benchmarks that have stayed the same:"
    titles["red"] = "Benchmarks that have got worse:"
    titles["lightgrey"] = "Benchmarks that are not comparable:"
    titles["all"] = "All benchmarks:"

    log.flush()

    for key in keys:
        if len(bench[key]) == 0:
            continue

        if not only_changed:
            color_print("")
            color_print(titles[key])
            color_print("")

        name_1 = False  # commit_names.get(hash_1)
        if name_1:
            name_1 = f"<{name_1}>"
        else:
            name_1 = ""

        name_2 = False  # commit_names.get(hash_2)
        if name_2:
            name_2 = f"<{name_2}>"
        else:
            name_2 = ""

        if sort == "default":
            pass
        elif sort == "ratio":
            bench[key].sort(key=lambda v: v[3], reverse=True)
        elif sort == "name":
            bench[key].sort(key=lambda v: v[2])
        else:
            raise ValueError("Unknown 'sort'")

        print(worsened, improved)
        return tabulate.tabulate(
            bench[key],
            headers=[
                "Change",
                f"Before {name_1}",
                f"After {name_2}",
                "Ratio",
                "Benchmark (Parameter)",
            ],
            tablefmt="github",
        )
