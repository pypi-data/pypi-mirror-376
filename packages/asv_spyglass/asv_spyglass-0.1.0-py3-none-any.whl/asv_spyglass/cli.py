from pathlib import Path

import click
import polars as pl
from asv import results

from asv_spyglass._asv_ro import ReadOnlyASVBenchmarks
from asv_spyglass.compare import ResultPreparer, do_compare


@click.group()
def cli():
    """
    Command-line interface for ASV benchmark analysis.
    """
    pass


@cli.command()
@click.argument("b1", type=click.Path(exists=True), required=True)
@click.argument("b2", type=click.Path(exists=True), required=True)
@click.argument("bconf", type=click.Path(exists=True), required=False)
@click.option(
    "--factor",
    default=1.1,
    show_default=True,
    help="Factor for determining significant changes.",
)
@click.option(
    "--split",
    is_flag=True,
    help="Split output by improvement/regression.",
)
@click.option(
    "--only-changed",
    is_flag=True,
    help="Only show changed benchmarks.",
)
@click.option(
    "--sort",
    type=click.Choice(["default", "ratio", "name"]),
    default="default",
    show_default=True,
    help="Sort output by change, ratio, or name.",
)
def compare(b1, b2, bconf, factor, split, only_changed, sort):  # Renamed to 'compare'
    """
    Compare two ASV result files.
    """
    if not bconf:
        bconf_path = Path(b1).parent.parent / "benchmarks.json"
        if bconf_path.exists():
            bconf = str(bconf_path)
        else:
            raise click.UsageError(
                "Error: Missing argument 'BCONF'. Could not find 'benchmarks.json' automatically. "
                "Please provide the path to your benchmarks.json file."
            )

    print(do_compare(b1, b2, bconf, factor, split, only_changed, sort))


@cli.command()
@click.argument("bres", type=click.Path(exists=True), required=True)
@click.argument("bdat", type=click.Path(exists=True), required=True)
@click.option(
    "--csv",
    type=click.Path(),
    help="Save data to csv",
)
def to_df(bres, bdat, csv):
    """
    Generate a dataframe from an ASV result file.
    """
    res = results.Results.load(bres)
    benchdat = ReadOnlyASVBenchmarks(Path(bdat)).benchmarks
    preparer = ResultPreparer(benchdat)
    df = preparer.prepare(res).to_df()
    if csv:
        df.write_csv(csv)
    else:
        with pl.Config(
            tbl_formatting="ASCII_MARKDOWN",
            tbl_hide_column_data_types=True,
            fmt_str_lengths=50,
            tbl_cols=50,
        ):
            click.echo(df)
