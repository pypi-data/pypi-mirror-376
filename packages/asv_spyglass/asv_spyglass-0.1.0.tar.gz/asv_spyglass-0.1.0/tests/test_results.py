import pprint as pp

from approvaltests.approvals import verify
from asv import results

from asv_spyglass._asv_ro import ReadOnlyASVBenchmarks
from asv_spyglass._aux import getstrform
from asv_spyglass.compare import (
    ResultPreparer,
    do_compare,
    result_iter,
)


def test_result_iter(shared_datadir):
    res = results.Results.load(
        getstrform(shared_datadir / "a0f29428-conda-py3.11-numpy.json")
    )
    verify(pp.pformat([tuple(x) for x in (result_iter(res))]))


def test_do_compare(shared_datadir):
    verify(
        do_compare(
            getstrform(shared_datadir / "a0f29428-conda-py3.11-numpy.json"),
            getstrform(shared_datadir / "a0f29428-virtualenv-py3.12-numpy.json"),
            shared_datadir / "asv_samples_a0f29428_benchmarks.json",
        )
    )


def test_result_df(shared_datadir):
    res = results.Results.load(
        getstrform(shared_datadir / "d6b286b8-rattler-py3.12-numpy.json")
    )
    benchdat = ReadOnlyASVBenchmarks(
        shared_datadir / "d6b286b8_asv_samples_benchmarks.json"
    ).benchmarks
    preparer = ResultPreparer(benchdat)
    pres1 = preparer.prepare(res).to_df()
    verify(pp.pformat(pres1.to_dict()))
