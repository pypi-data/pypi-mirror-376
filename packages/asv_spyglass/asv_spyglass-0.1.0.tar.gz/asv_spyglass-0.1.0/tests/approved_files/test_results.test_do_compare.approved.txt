| Change   | Before    | After      |   Ratio | Benchmark (Parameter)                                                                                 |
|----------|-----------|------------|---------|-------------------------------------------------------------------------------------------------------|
| -        | 94.8±30μs | 28.4±0.2μs |     0.3 | benchmarks.TimeSuite.time_add_arr [rgx1gen11/conda-py3.11-numpy -> rgx1gen11/virtualenv-py3.12-numpy] |
